Add preciousblock RPC

sipa opened the pull request:
This adds a new PRC 'preciousblock' to mark a block as precious. A precious block will be treated as if it was received earlier than any other competing block.

sipa commented:
This was suggested in #6995.

kanoi commented:
Awesome name :)

gmaxwell commented:
Concept ACK. 

This needs to document that overriding the first seen block behavior in slows network convergence and can increase the rate of long reorganizations.

MarcoFalke commented on src/rpcblockchain.cpp:
```diff
@@ -800,6 +800,44 @@ UniValue getmempoolinfo(const UniValue& params, bool fHelp)
+        throw JSONRPCError(RPC_DATABASE_ERROR, state.GetRejectReason());
+    }
+
+    return NullUniValue;
```
Nit: What about being more verbose and return true? Considering the UX part of that, some users may interpret an empty line as "failed".

<sub>Further reading: http://ux.stackexchange.com/a/12819</sub>

sipa replied:
It's very common to return Null in case there is no real result in our API.

MarcoFalke replied:
Some places also use `return true`. But yes, null is more common.

MarcoFalke replied:
new issue: #7043

TheBlueMatt commented:
I would prefer if it would only prefer the given block over other blocks that arrived within, eg, 10 seconds of the other block in question.

On November 12, 2015 3:59:35 AM PST, Pieter Wuille <notifications@github.com> wrote:
>This adds a new PRC 'preciousblock' to mark a block as precious. A
>precious block will be treated as if it was received earlier than any
>other competing block.
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/bitcoin/bitcoin/pull/6996
>
>-- Commit Summary --
>
>  * Add preciousblock RPC
>
>-- File Changes --
>
>    M qa/pull-tester/rpc-tests.py (1)
>    A qa/rpc-tests/preciousblock.py (66)
>    M src/chain.h (2)
>    M src/main.cpp (15)
>    M src/main.h (3)
>    M src/rpcblockchain.cpp (38)
>
>-- Patch Links --
>
>https://github.com/bitcoin/bitcoin/pull/6996.patch
>https://github.com/bitcoin/bitcoin/pull/6996.diff
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/6996

sipa commented:
@TheBlueMatt That would be a much larger change, as we don't track receive time.

TheBlueMatt commented:
Indeed, not sure it's worth it, but would prefer if we could :).

On November 12, 2015 12:33:08 PM PST, Pieter Wuille <notifications@github.com> wrote:
>@TheBlueMatt That would be a much larger change, as we don't track
>receive time.
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/6996#issuecomment-156225786

MarcoFalke commented:
utACK

gmaxwell commented:
I'm sympathetic with matt's ask; but I don't think it's worth it-- it'll get bypassed (which then puts parties back into a mode of modifying the consensus code), and it would imply that 10 seconds was "safe"-- I have no idea if it is... the inherent preference for your own blocks by virtue of hearing the first already hurts convergence; so I think it would be fair to say that no amount additional is "safe"... but having a well constructed option is much safer than people trying to achieve this with invalidateblock.

kanoi commented:
Well any numerical limitation would of course only require a few bytes of code change to circumvent so would seem to be somewhat a waste of time considering the extra work involved: as @sipa suggests

sipa commented:
@gmaxwell What do you mean by 'saturate' ?

sipa commented:
Ok, changed the approach a bit. It now only keeps a counter per tip, and resets it when more work was added.

gmaxwell commented:
By saturate I mean what you did,  sat_add(int_min, -1) = int_min

sipa commented:
Rebased.

jgarzik commented:
ut ACK

sipa commented:
There is a bug in the unit tests that can lead to a race condition.

sipa commented:
Fixed.

sipa commented:
Rebased.

sdaftuar commented on qa/rpc-tests/preciousblock.py:
```diff
@@ -0,0 +1,87 @@
+        assert(hashC != hashG)
+        print "Connect nodes and check no reorg occurs"
+        connect_nodes_bi(self.nodes,0,1)
+        sync_blocks(self.nodes[0:2])
```
FYI, i don't think this call to `sync_blocks` is doing anything helpful -- it is just checking that `getblockcount` is the same between the two nodes, which it will be immediately, even before any headers or blocks are sent between the two nodes.

Perhaps this should sync on `getchaintips()` returning the expected output?

sdaftuar commented on qa/rpc-tests/preciousblock.py:
```diff
@@ -0,0 +1,87 @@
+        self.nodes[0].preciousblock(hashC)
+        assert(self.nodes[0].getbestblockhash() == hashC)
+        print "Make Node1 prefer block C"
+        self.nodes[1].preciousblock(hashC)
```
There's a race condition here; it's possible for nodes[1] to have not yet received hashC at the time this preciousblock call happens, which will cause the sync_chain in the next line to hang indefinitely.

laanwj commented:
Needs rebase

luke-jr commented on qa/rpc-tests/preciousblock.py:
```diff
@@ -0,0 +1,87 @@
+
+    def setup_network(self):
+        self.nodes = []
+        self.is_network_split = False 
```
Extra space at the end of the line.

luke-jr commented:
This can break CheckBlockIndex:

```C++
if (pindex->nChainTx == 0) assert(pindex->nSequenceId == 0);  // nSequenceId can't be set for blocks that aren't linked
```

This assumption is no longer true if we use preciousblock on a not-yet-downloaded block (and this can occur in the RPC test).

gmaxwell commented:
Needs rebase.

arowser commented:
Can one of the admins verify this patch?

MarcoFalke commented on qa/rpc-tests/preciousblock.py:
```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python2
```
Needs to be changed to python3 after rebase.

Also some other adjustments are needed in this file.

btcdrak commented:
needs rebase

luke-jr commented:
Pushed a rebase to my repo

laanwj commented:
Are you still planning to do this?

MarcoFalke commented:
@sipa Any eta here? (I can take care of the tests, but this needs rebase)

sipa commented:
@luke-jr Your rebase does not fix the comment you made here: https://github.com/bitcoin/bitcoin/pull/6996#issuecomment-189314423 ?

sipa commented:
@luke-jr Switched to your commit e7a3457, and changed the test to python3 and updated the assert. The tests don't run, though. I'll inventigate later.

MarcoFalke commented on qa/rpc-tests/preciousblock.py:
```diff
@@ -0,0 +1,87 @@
+from test_framework.util import *
+
+class PreciousTest(BitcoinTestFramework):
+    def setup_chain(self):
```
I think you need to use `__init__()` here, instead.

luke-jr commented:
@sipa It is fixed in eafeb97f971e452720aeb645175384f660393660, and the tests updated in 5e6af8248f3980234e4aeb15f77ca61ed4124ef1 (both of which are part of my "preciousblock" branch)

sipa committed:
Add preciousblock RPC

Includes a bugfix by Luke-Jr.

sipa committed:
Add preciousblock tests

Rebased, improved and extended by Luke-Jr.

sipa commented:
@luke-jr Ah, a link to that would have been useful :)

I've squashed the fixes and split into 2 commits.

laanwj commented:
utACK 5805ac8

btcdrak commented:
utACK 5805ac836c5847bc54cbef3e71154d022ca18eda

rebroad commented:
that was a surprise merge! Are there any plans that anyone knows of to use this? I would love to see some use-cases for documentation purposes.
